docs/algorithms.md

```diff
@@ -32,9 +32,49 @@ Each course is associated with one average rating and one average workload, acce
 - Each course can have several questions tagged with `Overall` or `Workload`. This usually comes from the course having both a YC and a GS cross-listing. Each question should have 5 options each corresponding to its number of responses. We sum the response numbers across all questions.
 - Finally, we take the average: with 1, 2, 3, 4, 5 as the scores and the number of responses as the weights.
 
-## Course history
+## Same courses
 
-TODO
+Finding identical offerings of a class over time is essential to the following features:
+
+1. The historical mean rating and workload for a class, calculated over all past offerings of the same class.
+2. The most recent offering of a class, and its associated attributes (enrollment in particular).
+3. Any statistical analysis on how courses may have changed over time.
+
+It is one of our key algorithms that we spend the most time on.
+
+### Goals
+
+Our course matching algorithm should simultaneously avoid false-positives and false-negatives. While both are inevitable because the degree to which a class can change is on a continuum, in general we try to err on the generous side and prefer false-positives to false-negatives (that is, consider more courses to be the same course).
+
+In particular, here are a few methods that _don't work_:
+
+- For each course, list all courses with the same code as the same course. This creates too many false-positives, becuase course codes may be recycled. For example, CPSC 427 had been "Object-Oriented Programming" up to fall 2019, but changed to CPSC 327 in fall 2020; in fall 2024, CPSC 427 re-appears with the title "C++ Programming for Stability, Security, and Speed". We don't want to consider the two courses called CPSC 427 as the same course and link their ratings.
+- For each course, list all courses with the same title as the same course. This creates too many false-negatives, because course titles may change, such as changing capitalization, adding/removing words, etc. This may also create false-positives, for example in fall 2020, two courses both called "Foundations of Accounting and Valuation" coexisted in the catalog, but with completely different schools, professors, and course codes (ACCT 270 vs. MGT 502/HPM 502).
+- For each course, _only_ consider courses with the same code and similar title/description. This is generally fine! We can still successfully link two courses with unrelated codes together, as long as they use a "bridge" that cross-list both. However, consider the following hypothetical case: in season 1, course X has code A; in season 2, it has code A and B; in season 3, it has code B. It turns out that seasons 1 and 3 offered the same content but season 2 was different. We will fail to link them because the bridge doesn't exist. This may happen with seminars, where every season a different set of topics are offered.
```

Comment:
```suggestion
- For each course, _only_ consider courses with the same code and similar title/description. This is generally fine! We can still successfully link two courses with unrelated codes together, as long as they use a "bridge" that cross-list both (because we use a connected-components model; see below). However, consider the following hypothetical case: in season 1, course X has code A; in season 2, it has code A and B; in season 3, it has code B. It turns out that seasons 1 and 3 offered the same content but season 2 was different. We will fail to link them because the bridge doesn't exist. This may happen with seminars, where every season a different set of topics are offered.
```